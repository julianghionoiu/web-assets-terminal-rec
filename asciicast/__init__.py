import json
from decimal import Decimal

from colorama import Fore, Style


class Text:
    NEWLINE = "\r\n"
    CONSOLE_ARROW = Fore.CYAN + "âžœ " + Style.RESET_ALL


def wait(ticks):
    return [{"duration_in_ticks": ticks, "content": ""}]


def appear(content):
    return [{"duration_in_ticks": 1, "content": content}]


def typing(text="hostname"):
    return [{"duration_in_ticks": 1, "content": char} for char in text]


def delete(num_chars):
    return [{"duration_in_ticks": 1, "content": "\b \b"} for _ in range(0, num_chars)]


def press_enter():
    return [{"duration_in_ticks": 1, "content": Text.NEWLINE}]


def red(text):
    return Fore.RED + text + Style.RESET_ALL


def green(text):
    return Fore.GREEN + text + Style.RESET_ALL


def dim(text):
    return Style.DIM + text + Style.RESET_ALL


def bright(text):
    return Style.BRIGHT + text + Style.RESET_ALL


class DecimalEncoder(json.JSONEncoder):
    def default(self, obj):
        if isinstance(obj, Decimal):
            return float("{:.2f}".format(obj))
        # Let the base class default method raise the TypeError
        return json.JSONEncoder.default(self, obj)


def write_asciicast_v2(stream, metadata, asciicast_lines):
    length_of_one_tick_in_seconds = 0.10
    current_time_sec = Decimal(0.00)
    json.dump(metadata, stream)
    stream.write("\n")
    for line in asciicast_lines:
        asciicast_v2_line = [current_time_sec, "o", line["content"]]
        current_time_sec += Decimal(length_of_one_tick_in_seconds * line["duration_in_ticks"])

        json.dump(asciicast_v2_line, stream, cls=DecimalEncoder)
        stream.write("\n")
