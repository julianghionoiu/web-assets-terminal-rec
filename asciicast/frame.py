class AsciicastFrame(object):
    duration_in_ticks: int
    content: str

    def __init__(self, duration_in_ticks: int, content: str) -> None:
        self.duration_in_ticks = duration_in_ticks
        self.content = content
