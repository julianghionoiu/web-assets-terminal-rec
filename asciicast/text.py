def console_arrow() -> str:
    return "\u001b[36mâžœ \u001b[m"


def thumbs_up() -> str:
    # Find code here: https://unicode.org/emoji/charts/full-emoji-list.html#1f44d
    # Get Java reference here: http://www.fileformat.info/info/unicode/char/1f44d/index.htm
    return "\uD83D\uDC4D"


def newline() -> str:
    return "\r\n"


# ~~~~~~~~~~ Colours

def black(text: str) -> str:
    return _color(30, text)


def red(text: str) -> str:
    return _color(31, text)


def green(text: str) -> str:
    return _color(32, text)


def yellow(text: str) -> str:
    return _color(33, text)


def blue(text: str) -> str:
    return _color(34, text)


def magenta(text: str) -> str:
    return _color(35, text)


def cyan(text: str) -> str:
    return _color(36, text)


def white(text: str) -> str:
    return _color(37, text)


def bright(text: str) -> str:
    return _color(1, text)


def dim(text: str) -> str:
    return _color(2, text)


def _color(code: int, text: str):
    return "\u001b[{}m{}\u001b[0m".format(code, text)


# ~~~~~~~~~~ Utils

def trim(text: str, length: int):
    return (text[:20] + '..') if len(text) > length else text


def replace_escapes(text: str):
    return text.replace("\u001b", "Esc")
