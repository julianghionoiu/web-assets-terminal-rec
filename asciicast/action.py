from abc import ABC, abstractmethod
from typing import List

from asciicast.frame import AsciicastFrame


class AsciicastAction(ABC):

    @abstractmethod
    def render_frames(self, width: int, height: int) -> List[AsciicastFrame]:
        ...
