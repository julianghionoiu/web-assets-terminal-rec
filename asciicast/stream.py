import json
from decimal import Decimal

from asciicast.frame import AsciicastFrame


class AsciicastStream(object):
    width: int
    height: int

    def __init__(self, width: int, height: int, stream, length_of_one_tick_in_seconds=0.10) -> None:
        self.width = width
        self.height = height
        self._length_of_one_tick_in_seconds = length_of_one_tick_in_seconds

        # Start the stream
        self._stream = stream
        self._current_time_sec = Decimal(0.00)
        metadata = {"version": 2, "width": self.width, "height": self.height}
        json.dump(metadata, stream)
        stream.write("\n")

    def write_frame(self, frame: AsciicastFrame):
        asciicast_v2_line = [self._current_time_sec, frame.frame_type, frame.content]
        self._current_time_sec += Decimal(self._length_of_one_tick_in_seconds * frame.duration_in_ticks)
        json.dump(asciicast_v2_line, self._stream, cls=DecimalEncoder)
        self._stream.write("\n")


class DecimalEncoder(json.JSONEncoder):
    def default(self, obj):
        if isinstance(obj, Decimal):
            return float("{:.2f}".format(obj))
        # Let the base class default method raise the TypeError
        return json.JSONEncoder.default(self, obj)
