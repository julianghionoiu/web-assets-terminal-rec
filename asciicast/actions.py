from typing import List

from asciicast.action import AsciicastAction
from asciicast.frame import AsciicastFrame
from asciicast.text import newline


class Wait(AsciicastAction):
    _ticks: int

    def __init__(self, ticks: int) -> None:
        self._ticks = ticks

    def render_frames(self, width: int, height: int) -> List[AsciicastFrame]:
        return [AsciicastFrame(self._ticks, "")]


class Appear(AsciicastAction):
    _content: str

    def __init__(self, content: str) -> None:
        self._content = content

    def render_frames(self, width: int, height: int) -> List[AsciicastFrame]:
        return [AsciicastFrame(1, self._content)]


class TypeChars(AsciicastAction):
    _text: str

    def __init__(self, text: str) -> None:
        self._text = text

    def render_frames(self, width: int, height: int) -> List[AsciicastFrame]:
        return [AsciicastFrame(1, char) for char in self._text]


class Delete(AsciicastAction):
    _num_chars: int

    def __init__(self, num_chars: int) -> None:
        self._num_chars = num_chars

    def render_frames(self, width: int, height: int) -> List[AsciicastFrame]:
        return [AsciicastFrame(1, "\b \b") for _ in range(0, self._num_chars)]


class PressEnter(AsciicastAction):

    def render_frames(self, width: int, height: int) -> List[AsciicastFrame]:
        return [AsciicastFrame(1, newline())]
