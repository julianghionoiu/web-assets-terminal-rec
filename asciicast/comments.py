from asciicast.frame import AsciicastFrame, AsciicastFrameType
from asciicast.stream import AsciicastStream


class Comments(object):

    def __init__(self, asciicast_stream: AsciicastStream) -> None:
        self._asciicast_stream = asciicast_stream

    def add(self, comment: str) -> None:
        self._asciicast_stream.write_frame(
            AsciicastFrame(0,
                           frame_type=AsciicastFrameType.INPUT,
                           content=comment))
