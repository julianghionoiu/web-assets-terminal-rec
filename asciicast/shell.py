from asciicast.stream import AsciicastStream
from asciicast.text import newline


class Shell(object):

    def __init__(self, asciicast_stream: AsciicastStream) -> None:
        self._asciicast_stream = asciicast_stream

    def wait(self, ticks: int) -> None:
        self._asciicast_stream.write_frame(ticks, content="")

    def appear(self, content: str) -> None:
        self._asciicast_stream.write_frame(1, content=content)

    def type_chars(self, text: str) -> None:
        for _char in text:
            self._asciicast_stream.write_frame(1, content=_char)

    def delete(self, num_chars: int) -> None:
        for _ in range(0, num_chars):
            self._asciicast_stream.write_frame(1, content="\b \b")

    def press_enter(self):
        self._asciicast_stream.write_frame(1, content=newline())
