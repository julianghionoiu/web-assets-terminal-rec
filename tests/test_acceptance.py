import io

from approvaltests.approvals import verify

from asciicast import appear, Text, write_asciicast_v2, red, green, bright, dim, delete
from asciicast import wait, typing, press_enter


def test_appear_and_type():
    metadata = {"version": 2, "width": 60, "height": 21}

    asciicast_lines = []
    asciicast_lines += wait(ticks=10)
    asciicast_lines += appear("> ")
    asciicast_lines += typing("hostname")
    asciicast_lines += press_enter()
    asciicast_lines += appear("candidate-laptop.local" + Text.NEWLINE)

    asciicast_v2_file = io.StringIO()
    write_asciicast_v2(asciicast_v2_file, metadata, asciicast_lines)

    verify(asciicast_v2_file.getvalue())


def test_delete():
    metadata = {"version": 2, "width": 60, "height": 21}

    asciicast_lines = []

    asciicast_lines += appear("> ")
    asciicast_lines += appear("foocar")
    asciicast_lines += delete(3)
    asciicast_lines += appear("bar")

    asciicast_v2_file = io.StringIO()
    write_asciicast_v2(asciicast_v2_file, metadata, asciicast_lines)

    verify(asciicast_v2_file.getvalue())


def test_colours():
    metadata = {"version": 2, "width": 60, "height": 21}

    asciicast_lines = []

    asciicast_lines += appear("1. " + red("some red text") + " " + green('some green text') + Text.NEWLINE)
    asciicast_lines += appear("2. " + dim(red('x')) + " " + bright(red('x')) + Text.NEWLINE)

    asciicast_v2_file = io.StringIO()
    write_asciicast_v2(asciicast_v2_file, metadata, asciicast_lines)

    verify(asciicast_v2_file.getvalue())
