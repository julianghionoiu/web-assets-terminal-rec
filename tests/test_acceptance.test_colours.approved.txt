{"version": 2, "width": 60, "height": 21}
[0.0, "o", "1. \u001b[31msome red text\u001b[0m \u001b[32msome green text\u001b[0m\r\n"]
[0.1, "o", "2. \u001b[2m\u001b[31mx\u001b[0m\u001b[0m \u001b[1m\u001b[31mx\u001b[0m\u001b[0m\r\n"]
